sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala

```diff
@@ -402,6 +403,22 @@ case class PreprocessTableInsertion(conf: SQLConf) extends Rule[LogicalPlan] {
           s"including ${staticPartCols.size} partition column(s) having constant value(s).")
     }
 
+    val partitionsTrackedByCatalog = conf.manageFilesourcePartitions &&
+      catalogTable.isDefined &&
+      catalogTable.get.partitionColumnNames.nonEmpty &&
+      catalogTable.get.tracksPartitionsInCatalog
+    // check static partition
+    if (partitionsTrackedByCatalog &&
+      normalizedPartSpec.nonEmpty &&
+      staticPartCols.size == partColNames.size) {
```

Comment:
why do we need this check `staticPartCols.size == partColNames.size`?

Reply:
`staticPartCols` only has fields that specify the partition value
`partColNames` are all partition fields
Because it only checks the writing to the static partition

```sql
insert ovwriter table a partition(d='1',h) select 1,''
```
```
staticPartCols=[d]
partColNames=[d,h]
```
```sql
insert ovwriter table a partition(d='1',h='') select 1
```
```
staticPartCols=[d,h]
partColNames=[d,h]
```

Reply:
`partition(d='',h)` should we fail for this case?

Reply:
Spark/Hive allows the partition value to be empty for dynamic partition write.


```
Loading data to table x.t1 partition (d=1, h=null)
	 Time taken for load dynamic partitions : 203
	Loading partition {d=1, h=__HIVE_DEFAULT_PARTITION__}
	 Time taken for adding to write entity : 1
Partition x.t1{d=1, h=__HIVE_DEFAULT_PARTITION__} stats: [numFiles=1, numRows=1, totalSize=201, rawDataSize=85]
```

Reply:
your example is `partition(d='1',h)`, I was asking about `partition(d='',h)`

Reply:
then shall we check all `staticPartCols` have partition values?

Reply:
Yes, we can do this check, but it will cause errors in such SQL execution, because it was successful before.

Reply:
but the behavior is inconsistent. It's weird to see `partition(d='',h)` works, but `partition(d='',h='1')` does not.

Reply:
make sense.